### PR TITLE
bugfix: clean root context when state machine inst record failed

### LINKED ISSUE #3312: bug: SAGA模式创建状态机实例，在存入数据库时，如果出现异常，会导致xid和branchType未从RootContext中移除，全局事务也未处理掉。
- [ ] I have searched the [issues](https://github.com/seata/seata/issues) of this repository and believe that this is not a duplicate.

### Ⅰ. Issue Description
bug: SAGA模式在创建状态机实例，在存入数据库时，如果出现异常，会导致xid和branchType未从RootContext中移除。
同时，全局事务数据也没有被处理掉，只能等待超时回滚。

图片

### Ⅳ. How to reproduce it (as minimally and precisely as possible)

1. 在执行保存状态机实例数据时，故意将client与数据库网络中断掉，即可重现该问题。

### Ⅵ. Environment:

- JDK version : 1.8
- OS : windows10
- Others: seata1.4.0

### PR DESCRIPTION
**Pull request by anselleeyy**
### Ⅰ. Describe what this PR did

fixbug: SAGA模式在创建状态机实例，在存入数据库时，出现异常时，移除 xid和branchType，避免影响其他事务执行

### Ⅱ. Does this pull request fix one issue?

fixed #3312

### CODE REVIEW

**Comment by anselleeyy** [comment]
@long187 fixed, use duplicate key exception instead

**Review by wangliang181230 — APPROVED** [review]
LGTM

File: test/src/test/java/io/seata/saga/engine/db/StateMachineDBExceptionTests.java [hunk 1]
```diff
@@ -0,0 +1,61 @@
...
+        ApplicationContext applicationContext = new ClassPathXmlApplicationContext("classpath:saga/spring/statemachine_engine_db_test.xml");
+        stateMachineEngine = applicationContext.getBean("stateMachineEngine", StateMachineEngine.class);
+        jdbcTemplate = applicationContext.getBean(JdbcTemplate.class);
+    }
+
+    @Test
+    public void testStateMachineRecordFailed() {
+
+        // drop table to mock record exception
+        jdbcTemplate.execute("drop table seata_state_machine_inst");
```
[the remark this answers is not in the document]
Comment (anselleeyy) [reply test/src/test/java/io/seata/saga/engine/db/StateMachineDBExceptionTests.java]: fixed, use duplicate key exception instead

**Review by long187 — APPROVED** [review]
LTGM

**Review by slievrly — APPROVED** [review]
LGTM

**Pull request merged**